webapp/packages/plugin-data-spreadsheet-new/src/DataGrid/useTableData.tsx

```diff
@@ -47,22 +48,120 @@ const COLUMN_HEADER_ICON_WIDTH = 16;
 const COLUMN_HEADER_TEXT_PADDING = 8;
 const COLUMN_HEADER_ORDER_PADDING = 8;
 const COLUMN_HEADER_ORDER_WIDTH = 16;
+const COLUMNS_PER_CHUNK = 10;
+const MIN_COLUMN_WIDTH = 300;
 
 const FONT = '400 12px Roboto';
 
 export function useTableData(
   model: IDatabaseDataModel<any, IDatabaseResultSet>,
   resultIndex: number,
   gridDIVElement: React.RefObject<HTMLDivElement | null>,
-): ITableData {
+): [ITableData, ITableState] {
   const format = model.source.getAction(resultIndex, ResultSetFormatAction);
   const data = model.source.getAction(resultIndex, ResultSetDataAction);
   const editor = model.source.getAction(resultIndex, ResultSetEditAction);
   const view = model.source.getAction(resultIndex, ResultSetViewAction);
   const dataContent = model.source.getAction(resultIndex, ResultSetDataContentAction);
   const constraints = model.source.getAction(resultIndex, ResultSetConstraintAction);
+  const columnsRef = useRef<Column<IResultSetRowKey, any>[]>([]);
+  const timers = useRef<(NodeJS.Timeout | undefined)[]>([]);
 
-  return useObservableRef<ITableData & { gridDIVElement: React.RefObject<HTMLDivElement | null> }>(
+  const scheduleColumnsUpdate = (currentChunk = 0) => {
+    const totalChunksAmount = Math.ceil(format.getHeaders().length / COLUMNS_PER_CHUNK);
+    const columnNames = format.getHeaders();
+    const start = currentChunk * COLUMNS_PER_CHUNK;
+    const end = Math.min(COLUMNS_PER_CHUNK + start, columnNames.length);
+    const rowStrings = format.getLongestCells(start, end);
+
+    const { columnsWidth, cellsWidth } = getWidths(rowStrings);
+
+    const newColumns = columnsRef.current.map((column, index) => {
+      if (index < start || index >= end) {
+        return column;
+      }
+
+      return {
+        ...column,
+        width: Math.min(MIN_COLUMN_WIDTH, Math.max(columnsWidth[index], cellsWidth[index] ?? 0)),
+      };
+    });
+
+    columnsRef.current = newColumns;
+    setColumns(newColumns);
+
+    if (currentChunk < totalChunksAmount - 1) {
+      timers.current[currentChunk] = setTimeout(() => {
+        scheduleColumnsUpdate(currentChunk + 1);
+
+        if (timers.current[currentChunk]) {
+          clearTimeout(timers.current[currentChunk]);
+          timers.current[currentChunk] = undefined;
+        }
+      }, 0);
+    }
+  };
+
+  const getWidths = (rowStrings: string[]) => {
+    const columnNames = format.getHeaders();
+
+    const columnsWidth = TextTools.getWidth({
+      font: FONT,
+      text: columnNames,
+    }).map(
+      width =>
+        width + COLUMN_PADDING + COLUMN_HEADER_ICON_WIDTH + COLUMN_HEADER_TEXT_PADDING + COLUMN_HEADER_ORDER_PADDING + COLUMN_HEADER_ORDER_WIDTH,
+    );
+
+    const cellsWidth = TextTools.getWidth({
+      font: FONT,
+      text: rowStrings,
+    }).map(width => width + COLUMN_PADDING);
+
+    return {
+      columnsWidth,
+      cellsWidth,
+    };
+  };
+
+  const getColumns = () => {
```

Comment:
please don't use arrow functions for functions with logic (it's hard to debug anonymous functions)
good: `() => calcSomething()`
bad: `() => a + b`